Core/ALife.Core/WorldObjects/Agents/Brains/NeuralNetworkBrain.cs

```diff
@@ -28,217 +31,318 @@ public NeuralNetworkBrain(Agent self, List<int> layers)
-                val += modification;
-                val = ExtraMath.Clamp(val, -1, 1);
+                layerInfoArray[i + 1] = (byte)Layers[i].Neurons.Count;
             }
-            return val;
-        }
 
-        private Layer CreateActionLayer(Agent self, Layer aboveLayer)
-        {
-            List<ActionNeuron> actionNeurons = new List<ActionNeuron>();
-            foreach(ActionCluster ac in self.Actions.Values)
+            List<double> doubleInfo = new List<double>();
+            doubleInfo.Add(ModificationRate);
+            doubleInfo.Add(MutabilityRate);
+
+            List<double> biases = new List<double>();
+            for(int j = 0; j < layerCount; ++j)
             {
-                foreach(ActionPart ap in ac.SubActions.Values)
-                {
-                    ActionNeuron neu = new ActionNeuron(ap);
-                    actionNeurons.Add(neu);
-                    for(int d = 0; d < aboveLayer.Neurons.Count; d++)
-                    {
-                        neu.UpstreamDendrites.Add(new Dendrite(aboveLayer.Neurons[d]));
-                    }
-                }
+                biases.AddRange(Layers[j].Neurons.Select(neu => neu.Bias));
             }
-            Layer actionLayer = new Layer(actionNeurons.Count);
-            actionLayer.Neurons.AddRange(actionNeurons);
-            return actionLayer;
-        }
+            //TODO: These could be added directly. But for now, easier to have some debuggability
+            doubleInfo.AddRange(biases);
 
-        private Layer CreateClonedActionLayer(Agent self, Layer aboveLayer, Layer templateLayer, bool exactCopy)
-        {
-            List<ActionNeuron> actionNeurons = new List<ActionNeuron>();
-            int apIndex = 0;
-            foreach(ActionCluster ac in self.Actions.Values)
+            List<double> dendriteWeights = new List<double>();
+            for(int k = 0; k < layerCount; ++k)
             {
-                foreach(ActionPart ap in ac.SubActions.Values)
+                Layer lay = Layers[k];
+                for(int j = 0; j < lay.Neurons.Count; ++j) 
                 {
-                    Neuron currNeuron = templateLayer.Neurons[apIndex];
-                    double neuBias = currNeuron.Bias;
-                    if(!exactCopy)
-                    {
-                        neuBias = ModifyBetweenNegOneAndOne(neuBias);
-                    }
-
-                    ActionNeuron neu = new ActionNeuron(ap, neuBias);
-                    actionNeurons.Add(neu);
-                    CreateDendrites(aboveLayer.Neurons, currNeuron, neu, exactCopy);
-                    apIndex++;
+                    Neuron n = lay.Neurons[j];
+                    dendriteWeights.AddRange(n.UpstreamDendrites.Select(den => den.Weight));
                 }
             }
-            if(actionNeurons.Count != templateLayer.Neurons.Count)
+            doubleInfo.AddRange(dendriteWeights);
+
+            List<string> NeuronNames = new List<string>();
```

Comment:
`NeuronNames` -> `neuronNames` so it is clearer further down it is a local variable.